### DISCUSSION TITLE
Preload data from CSV and refresh after Xmin.

### DISCUSSION THREAD

**Discussion by xhadev** [ben-manes/caffeine#898] [Q&A] [answered]
Hello,
I'm trying to work on solution based on CSV stored in a server( _I can't use a database_) that will hold the key/value data, that I need to load when my app start, that seems pretty straightforward with a `bulkload` approach but my problem is that after Xmin I want to refresh the whole cache a reload the CSV to fetch the latest values again, the CSV can change frequently. My problem is that the data is from a CSV so I don't have a single item load from a service/db other than my cache.

I know I can use a simple ConcurrentHashMap but I want to leverage the power off caffeine.

Note: I started my journey trying to use Guava, but _guavaAsyncLoader_ forces you to implement `load`
` public Boolean load(@Nonnull String key)` and I don't have a single load because the data is in the CSV. I just want that 
_.refreshAfterWrite(10, TimeUnit.MINUTES)_ will trigger a bulk load.

Thanks.

**Comment by ben-manes** [comment, marked as answer]
In these cases it is typically better to forgo a caching library and use a scheduled task to periodically replace an unmodifiable map. This is simple to reason about, works as desired, and requires no dependencies. For example,

```java
private volatile Map<String, String> cityToState;

CsvCache(ScheduledExecutorService executor) {
  executor.scheduleAtFixedRate(this::reload, 10, 10, TimeUnit.MINUTES);
  reload();
}

private reload() {
  cityToState = parseCsv().stream()
      .collect(toUnmodifiableMap(row -> row.get("city"), row -> row.get("state"));
}

public Optional<String> getStateForCity(String city) {
  return Optional.ofNullable(cityToState.get(city));
}
```
